File: backend/routes/complaints.py
from fastapi import APIRouter, UploadFile, File, Form
from fastapi.responses import JSONResponse
from database import SessionLocal
import models
import shutil
import os

router = APIRouter(prefix="/complaints", tags=["Complaints"])

UPLOAD_DIR = "uploads"
os.makedirs(UPLOAD_DIR, exist_ok=True)

SEED_COMPLAINTS = [
    {"name": "Rajesh Kumar",  "phone": "+91 98765 43210", "description": "Industrial waste being dumped directly into the river. Strong chemical smell and discoloration of water observed.", "location": "Sector 12, River Point A",   "severity": "high",   "media": "2 media file(s)", "status": "Pending"},
    {"name": "Priya Sharma",  "phone": "+91 87654 32109", "description": "Plastic waste and sewage overflow near residential area. Foul odor affecting nearby homes.",                        "location": "Sector 8, Bridge Area",        "severity": "medium", "media": "1 media file(s)", "status": "Verified"},
    {"name": "Amit Patel",    "phone": "+91 76543 21098", "description": "Foam and chemical effluents visible on river surface near the industrial outlet.",                                   "location": "Sector 15, Industrial Zone",  "severity": "high",   "media": "3 media file(s)", "status": "Resolved"},
    {"name": "Sunita Devi",   "phone": "+91 65432 10987", "description": "Garbage dumping on the riverbank affects the water quality during monsoon season.",                                  "location": "Sector 3, Residential Area",  "severity": "low",    "media": "1 media file(s)", "status": "Pending"},
    {"name": "Vikram Singh",  "phone": "+91 54321 09876", "description": "Unusual color change noticed in river water after heavy rains. Possible contamination from upstream.",               "location": "Sector 20, Downstream",       "severity": "medium", "media": "0 media file(s)", "status": "Rejected"},
]


def _seed(db):
    if db.query(models.Complaint).count() == 0:
        for idx, c in enumerate(SEED_COMPLAINTS):
            complaint = models.Complaint(
                name=c["name"], phone=c["phone"], description=c["description"],
                location=c["location"], severity=c["severity"],
                media=c["media"], status=c["status"]
            )
            db.add(complaint)
        db.commit()


@router.post("/")
async def create_complaint(
    name: str = Form(...),
    phone: str = Form(""),
    description: str = Form(...),
    location: str = Form(...),
    severity: str = Form("medium"),
    file: UploadFile = File(None),
):
    db = SessionLocal()
    media_path = ""

    if file and file.filename and file.filename != "no-file.txt":
        file_path = f"{UPLOAD_DIR}/{file.filename}"
        with open(file_path, "wb") as buffer:
            shutil.copyfileobj(file.file, buffer)
        media_path = file_path

    # Generate ID like RMS-2026-XXXX
    count = db.query(models.Complaint).count() + 1
    rms_id = f"RMS-2026-{1234 + count}"

    complaint = models.Complaint(
        name=name, phone=phone, description=description,
        location=location, severity=severity,
        media=media_path or "0 media file(s)", status="Pending"
    )
    db.add(complaint)
    db.commit()
    db.refresh(complaint)
    return {"msg": "Complaint submitted successfully", "id": rms_id}


@router.get("/")
def get_complaints():
    db = SessionLocal()
    _seed(db)
    complaints = db.query(models.Complaint).all()
    result = []
    for idx, c in enumerate(complaints):
        result.append({
            "id": f"RMS-2026-{1234 + idx}",
            "name": c.name,
            "phone": c.phone,
            "description": c.description,
            "location": c.location,
            "severity": c.severity or "medium",
            "media": c.media or "0 media file(s)",
            "status": c.status or "Pending",
            "timestamp": str(c.created_at)[:16].replace("T", " ") if c.created_at else "—",
        })
    return result


@router.patch("/{complaint_id}/status")
def update_complaint_status(complaint_id: str, body: dict):
    db = SessionLocal()
    # Find by index (id is RMS-based)
    try:
        idx = int(complaint_id.split("-")[-1]) - 1234
        complaints = db.query(models.Complaint).all()
        if 0 <= idx < len(complaints):
            complaints[idx].status = body.get("status", "Pending")
            db.commit()
    except Exception:
        pass
    return {"msg": "Status updated"}


@router.get("/{complaint_id}")
def get_complaint(complaint_id: int):
    db = SessionLocal()
    return db.query(models.Complaint).filter(models.Complaint.id == complaint_id).first()

File: frontend/js/complaints.js
/* =============================================
   complaints.js — Admin complaints view
   ============================================= */

const DEMO_COMPLAINTS_DATA = [
  { id: "RMS-2026-1234", name: "Rajesh Kumar",  severity: "high",   status: "Pending",  location: "Sector 12, River Point A",   phone: "+91 98765 43210", description: "Industrial waste being dumped directly into the river. Strong chemical smell and discoloration of water observed.", media: "2 media file(s)", timestamp: "2026-04-01 08:24" },
  { id: "RMS-2026-1235", name: "Priya Sharma",  severity: "medium", status: "Verified", location: "Sector 8, Bridge Area",        phone: "+91 87654 32109", description: "Plastic waste and sewage overflow near residential area. Foul odor affecting nearby homes.", media: "1 media file(s)", timestamp: "2026-03-31 15:40" },
  { id: "RMS-2026-1236", name: "Amit Patel",    severity: "high",   status: "Resolved", location: "Sector 15, Industrial Zone",  phone: "+91 76543 21098", description: "Foam and chemical effluents visible on river surface near the industrial outlet. Already informed local authorities.", media: "3 media file(s)", timestamp: "2026-03-30 09:12" },
  { id: "RMS-2026-1237", name: "Sunita Devi",   severity: "low",    status: "Pending",  location: "Sector 3, Residential Area",  phone: "+91 65432 10987", description: "Garbage dumping on the riverbank affecting water quality during monsoon season.", media: "1 media file(s)", timestamp: "2026-03-29 11:55" },
  { id: "RMS-2026-1238", name: "Vikram Singh",  severity: "medium", status: "Rejected", location: "Sector 20, Downstream",       phone: "+91 54321 09876", description: "Unusual color change noticed in river water after heavy rains. Possible contamination from upstream.", media: "0 media file(s)", timestamp: "2026-03-28 17:30" },
  { id: "RMS-2026-1239", name: "Meena Gupta",   severity: "low",    status: "Verified", location: "Sector 5, Park Area",         phone: "+91 43210 98765", description: "Dead fish spotted floating near the park area. Suspected pesticide runoff from agricultural fields.", media: "2 media file(s)", timestamp: "2026-03-27 14:20" },
];

let allComplaints = [];
let currentFilter = "all";

async function loadComplaints() {
  try {
    const res = await fetch(`${API}/complaints/`);
    const data = await res.json();
    allComplaints = Array.isArray(data) && data.length > 0 ? data : DEMO_COMPLAINTS_DATA;
  } catch (_) {
    allComplaints = DEMO_COMPLAINTS_DATA;
  }
  renderComplaints();
}

function setFilter(status, btn) {
  currentFilter = status;
  document.querySelectorAll(".filter-tab").forEach(t => t.classList.remove("active"));
  btn.classList.add("active");
  renderComplaints();
}

function filterComplaints() { renderComplaints(); }

function renderComplaints() {
  const query = (document.getElementById("searchInput")?.value || "").toLowerCase();
  const list  = document.getElementById("complaintsList");
  const filtered = allComplaints.filter(c => {
    const matchStatus = currentFilter === "all" || c.status?.toLowerCase() === currentFilter.toLowerCase();
    const matchSearch = !query || (c.name || "").toLowerCase().includes(query) ||
      (c.location || "").toLowerCase().includes(query) || (c.id || "").toLowerCase().includes(query);
    return matchStatus && matchSearch;
  });

  list.innerHTML = "";
  if (filtered.length === 0) {
    list.innerHTML = `<div style="text-align:center;color:var(--text-muted);padding:40px 0;font-size:14px;">No complaints found</div>`;
    return;
  }

  filtered.forEach((c, idx) => {
    const card = document.createElement("div");
    card.className = "complaint-card";
    card.innerHTML = `
      <div class="complaint-card-header">
        <div>
          <div class="complainant-name">${c.name || "Unknown"} <span class="badge ${(c.severity || 'low').toLowerCase()}">${c.severity || 'low'}</span></div>
          <div class="complaint-location">${c.location || ""}</div>
          <div class="complaint-id">${c.id || ""}</div>
        </div>
        <span class="badge ${(c.status || 'pending').toLowerCase()}">⏳ ${c.status || "Pending"}</span>
      </div>
      <div class="complaint-desc">${c.description || ""}</div>
      <div class="complaint-meta">${c.media ? `📎 ${c.media}` : ""}</div>
    `;
    card.onclick = () => showDetail(c, card);
    list.appendChild(card);
  });
}

function showDetail(c, card) {
  document.querySelectorAll(".complaint-card").forEach(el => el.classList.remove("selected"));
  card.classList.add("selected");

  const panel = document.getElementById("detailPanel");
  panel.className = "detail-panel has-content";
  panel.innerHTML = `
    <div class="detail-header">
      <div class="detail-name">${c.name}</div>
      <div style="display:flex;gap:6px;margin-top:6px;">
        <span class="badge ${c.severity?.toLowerCase()}">${c.severity}</span>
        <span class="badge ${c.status?.toLowerCase()}">${c.status}</span>
      </div>
    </div>
    <div class="detail-section"><div class="ds-label">Complaint ID</div><div class="ds-value">${c.id || "—"}</div></div>
    <div class="detail-section"><div class="ds-label">Location</div><div class="ds-value">📍 ${c.location}</div></div>
    <div class="detail-section"><div class="ds-label">Phone</div><div class="ds-value">${c.phone || "—"}</div></div>
    <div class="detail-section"><div class="ds-label">Submitted</div><div class="ds-value">${c.timestamp || "—"}</div></div>
    <div class="detail-section"><div class="ds-label">Description</div><div class="ds-value" style="line-height:1.6;">${c.description}</div></div>
    <div class="detail-section"><div class="ds-label">Evidence</div><div class="ds-value">📎 ${c.media || "No files"}</div></div>
    <div class="detail-actions">
      <button class="btn btn-success" onclick="updateStatus('${c.id}', 'Verified')">✔ Verify</button>
      <button class="btn btn-primary" onclick="updateStatus('${c.id}', 'Resolved')">✅ Resolve</button>
      <button class="btn btn-danger"  onclick="updateStatus('${c.id}', 'Rejected')">✖ Reject</button>
    </div>
  `;
}

async function updateStatus(id, newStatus) {
  try {
    await fetch(`${API}/complaints/${id}/status`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ status: newStatus })
    });
  } catch (_) {}

  // Update local data
  const c = allComplaints.find(x => x.id === id);
  if (c) c.status = newStatus;

  showToast(`Status updated to ${newStatus}`);
  renderComplaints();

  // Clear detail panel
  const panel = document.getElementById("detailPanel");
  panel.className = "detail-panel";
  panel.innerHTML = `<div class="detail-empty"><div class="detail-icon">👁️</div><p>Select a complaint to view details</p></div>`;
}

loadComplaints();